### PR TITLE
Conversion path for TOSA-to-MIOpen

### PR DESCRIPTION
**Pull request by sjw36**
* bufferization conversion
* Add linalg.reshape ops for 5d memrefs (nhwcg)
* TODO: add bias with linalg generic

### CODE REVIEW

**Comment by sjw36** [comment]
Addresses https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/101

**Comment by whchung** [comment]
I want to understand the broader picture first. Right now this PR converts `tosa.conv2d` to `miopen.conv2d`. But I don't think it runs end-to-end yet.

How do we define the end-to-end lowering process where GPU kernels could be formed, with elementwise operations be embedded within `miopen.conv2d`-like operations?

File: mlir/test/Conversion/TosaToMIOpen/tosa-to-miopen.mlir [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// RUN: mlir-opt --split-input-file --tosa-to-miopen-on-tensors %s -verify-diagnostics -o -| FileCheck %s
+
+// CHECK-LABEL: test_fusion
+// CHECK: miopen.conv2d(%{{.*}}, %{{.*}}, %{{.*}}) {arch = {{.*}}, dilations = [1 : i32, 1 : i32], filter_layout = ["k", "y", "x", "c", "g"], input_layout = ["ni", "hi", "wi", "ci", "gi"], num_cu = {{.*}} : i32, output_layout = ["no", "ho", "wo", "ko", "go"], padding = [0 : i32, 0 : i32, 0 : i32, 0 : i32], strides = [1 : i32, 1 : i32]} : memref<128x8x3x3x1xf32>, memref<128x8x32x32x1xf32>, memref<128x128x30x30x1xf32>
+    
+func @test_fusion(%arg0: tensor<128x8x32x32xf32>, %arg1: tensor<128x8x3x3xf32>, %arg2: tensor<8xf32>) -> tensor<128x128x30x30xf32> {
+  %0 = "tosa.conv2d"(%arg0, %arg1, %arg2) {dilation = [1, 1], pad = [0, 0, 0, 0], stride = [1, 1]} : (tensor<128x8x32x32xf32>, tensor<128x8x3x3xf32>, tensor<8xf32>) -> tensor<128x128x30x30xf32>
```
Comment (jerryyin) [inline mlir/test/Conversion/TosaToMIOpen/tosa-to-miopen.mlir:7]: [Question] Shall we need to provide client with this interface in Miir library sometime in the future?

Comment (sjw36) [reply mlir/test/Conversion/TosaToMIOpen/tosa-to-miopen.mlir:7]: Possibly if we want to optimize full or partial graph. Client could pass an IR in and we could pass an IR back with substitutions (fully realized kernels). For MIGraphX we could read ONNX directly and translate to TOSA. Alternatively we could have an API similar to NNAPI but which we stored internally as TOSA. Still fleshing out the ideas..

Comment (jerryyin) [reply mlir/test/Conversion/TosaToMIOpen/tosa-to-miopen.mlir:7]: Sounds good. I take your response as this being an area where a lot of change can happen. Therefore, if we want any optimization/update in the interface, we should hold off. Until we have defined the way client can take advantage for full-graph optimization, it is premature to decide on the implementation details of the specific interface.

**Comment by sjw36** [comment]
> I want to understand the broader picture first. Right now this PR converts `tosa.conv2d` to `miopen.conv2d`. But I don't think it runs end-to-end yet.
> 
> How do we define the end-to-end lowering process where GPU kernels could be formed, with elementwise operations be embedded within `miopen.conv2d`-like operations?

I can lower through all miopen lowering steps but the linalg ops are left untouched. I think we will need to lower them based on the blockwise, then threadwise dimensions of the conv2d. For start we will partition only a single conv2d with adjacent element-wise ops (as in this example), so probably change our lowering steps to Function level passes and store the lowered dimensions on the parent function. I will write this up so we can discuss as soon as I finish some experiments.

**Comment by whchung** [comment]
@jungpark-mlir / @sjw36 how do we want to proceed for this PR?

**Comment by sjw36** [comment]
I am rebasing to TOT. If there are no objections it would be good to commit since there are a few people using and I have some follow-on PRs.

**Comment by whchung** [comment]
@sjw36 SGTM. Once you finish rebase please remove `skip-ci` label.

**Comment by sjw36** [comment]
@whchung please approve.

**Review by whchung — APPROVED** [review]
We went through the process with @sjw36 in an offline meeting. I'm approving this PR.

**Pull request merged**